pretyping/evarsolve.ml

```diff
@@ -1356,17 +1354,32 @@ let has_constrainable_free_vars env evd aliases force k ev (fv_rels,fv_ids,let_r
 
 exception EvarSolvedOnTheFly of evar_map * EConstr.constr
 
+let is_default = function
+| SList.Default (_, SList.Nil) | SList.Nil -> true
+| _ -> false
+
 (* Try to project evk1[argsv1] on evk2[argsv2], if [ev1] is a pattern on
    the common domain of definition *)
 let project_evar_on_evar force unify flags env evd aliases k2 pbty (evk1,argsv1 as ev1) (evk2,argsv2 as ev2) =
+  let evi1 = Evd.find evd evk1 in
+  let evi2 = Evd.find evd evk2 in
+  if Option.is_empty pbty && is_default argsv1 && is_default argsv2 &&
+    Option.is_empty evi1.evar_candidates && Option.is_empty evi2.evar_candidates &&
+    let l1 = SList.length argsv1 in
+    Int.equal l1 (SList.length argsv2) &&
+    Int.equal l1 (Range.length (Environ.named_context_val env).env_named_idx)
+  then
+    (* Fast path *)
```

Comment:
Rather than this comment, it would be better to write why the conditions above justify this path. I don't get the last one for example. It would also make sense to name the predicate, and put it where the data is, so that if one adds something crazy to evar info which would make the evar not "trivial" then he has a chance to fix the predicate.